### ISSUE TITLE
Idea: Macro Functions for Variable Definitions & MetaData

### ISSUE THREAD

**Issue by daniel-samson** [salesagility/SuiteCRM#783]
I think it would be an good idea if we used macro functions to define the fields in variable definitions (vardefs). I have noticed that most of the array fields are duplicated. Being able to just define the minimal information could improve productivity, reduce errors and make our code look cleaner.

```
$dictionary['Account']['fields'] = array(
        IDField(['name' => 'id']),
        DateField(['name' =>'date_entered']),
        TextField([
            'name' => 'name',
            'vname' => 'LBL_NAME',
        ])
    );
```

We could even use this to automate the values of the fields. For example the value of an ID Field could be automatically created via the macro like so:

```
function IDField($arg) {
    if(!array_key_exists('name',$arg)) {
        $arg['name'] = 'id';
    }
    $array = array(
        'name' => $arg['name'],
        'value' => create_guid(),
        'vname' => 'LBL_ID',
        'type' => 'id',
        'len' => 255,
        'required' => false,
        'reportable'=> false,
        'massupdate' => false,
    );
    $array = array_merge_recursive($array, $arg);
    $field = array($arg['name'] => $array);
    return $field;
}
```

if all the variable definitions were defined like this. It would have the potential of making it easier to effect the behaviour of the entire application. For example, if we wanted to add or change the array fields across SuiteCRM. Convert a field into a function field across the SuiteCRM

Theses macros also have the potential to be extended or customised. Depending on how we implement them. Link Fields and Relationship Fields could made easy to create. We could make the fields in the middle tables easier to produce.

This could also be used else where such as meta data: 

```
  array (
    AddressMetaField(['name' => 'primary_address_street']),
    AddressMetaField(['name' => 'alt_address_street']);
  )
```

instead of 

```
array (
      array (
          'name' => 'primary_address_street',
          'hideLabel' => true,      
          'type' => 'address',
          'displayParams'=>array('key'=>'primary', 'rows'=>2, 'cols'=>30, 'maxlength'=>150),
      ),
      array (
          'name' => 'alt_address_street',
          'hideLabel'=>true,
          'type' => 'address',
          'displayParams'=>array('key'=>'alt', 'copy'=>'primary', 'rows'=>2, 'cols'=>30, 'maxlength'=>150),      
      ),
    ),
```

Also new custom field types could be defined more easily. This method would still be backwards compatible as the macro functions just return the same result.

**Comment by samus-aran** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Idea: Macro Functions for Variable Definitions & MetaData](https://trello.com/c/rO80s95i/128-idea-macro-functions-for-variable-definitions-metadata)

**Comment by samus-aran** [comment]
We have now set up a new home for suggestions at Trello.  All github issues that were labeled 'suggestion' have been moved and will be closed.  Certain ones will be progressed within the new Suggestion Box and may be re-opened.

Announcement of moving Suggestions:
https://suitecrm.com/forum/suggestion-box/13691-moving-suggestions-from-github-to

New SuiteCRM Suggestion Box
https://trello.com/b/Ht7LbMqw/suitecrm-suggestion-box
